biocrnpyler/components_combinatorial_complex.py

```diff
@@ -0,0 +1,342 @@
+        #Compute Species that need to be added to s0 to get the Complex sf
+
+        if not isinstance(sf, ComplexSpecies):
+            raise ValueError(f"sf must be a ComplexSpecies. Recieved {sf}")
+
+        species_to_add = []
+        for s in sf.species_set:
+
+            if s0.monomer_eq(s): #this is used instead of == to deal with the potential for different parents
+                s0_count = 1
+            elif isinstance(s0, ComplexSpecies):
+                s0_count = s0.monomer_count(s)#This is used instead of .count to deal with species with different parents
+            else:
+                s0_count = 0
+            
+            #Add the correct number of s to the list
+            sf_count = sf.monomer_count(s)
+            if s0_count < sf_count:
+                species_to_add += (sf_count-s0_count)*[s]
+            elif s0_count > sf_count:
+                species_to_add = None #In this case, s0 contains more stuff than sf, so nothing should be returned
+                break
+            else:
+                pass #if they have the same number, do not add it
+
+        #s0 contains more or different species than sf, return None
+        if (not isinstance(s0, ComplexSpecies)) and (sf.monomer_count(s0) == 0):
+            species_to_add = None
+        elif isinstance(s0, ComplexSpecies) and any([s0.monomer_count(s) > sf.monomer_count(s) for s in s0.species_set]) and sf.monomer_count(s0) == 0:
+            species_to_add = None
+        elif len(species_to_add) == 0:
+            species_to_add = None
+
+        return species_to_add
+
+    def get_combinations_between(self, s0, sf):
+        """
+        Returns all combinations of Species to create the Complex sf from s0.
+        """
+
+        species_to_add = self.compute_species_to_add(s0, sf)
+        
+        if species_to_add is None or len(species_to_add) == 0:
+            return []
+        else:
+            #combinations (binder, bindee, complex_species) to be returned
+            combinations = []
+            #get all permutations of the species for different binding orders 
+            perms = permutations(species_to_add, len(species_to_add))
+
+            #Iterate through permutations
+            for perm in perms:
+                bindee = s0
+                for s in perm:
+                    binder = s
+                    if isinstance(bindee, ComplexSpecies):
+                        s_list = list(bindee.species)
+                    else:
+                        s_list = [bindee]
+
+                    s_list.append(s)
+                    cs = Complex(s_list)
+
+                    #append the new combination if it isn't excluded
+                    if binder not in self.excluded_states and bindee not in self.excluded_states and cs not in self.excluded_states:
+                        combinations.append((binder, bindee, cs))
+                    #update bindee
+                    bindee = cs
+
+            return combinations
```

Comment:
I don't want to think about how this can be applied to integrases at the moment but I'm sure that when I get around to that I'll have more comments

Reply:
I didn't think about that either.....but it might solve some of the binding enumeration problems if you integrate it into an IntegraseEnumerator in a clever way!